ramanujan/cmf/cmf.py

```diff
@@ -123,3 +137,68 @@ def limit(
             the walk multiplication as defined above.
         """
         return self.walk(trajectory, iterations, start).limit(vector)
+
+    def __getitem__(self, location):
+        """
+        Same as `potential`.
+        """
+        return self.potential(location)
+
+    def potential(self, location):
+        r"""
+        Returns the potential matrix of the conservative matrix field at a given location.
+
+        This function is not implemented using CMF.walk, to enable caching of CMF potential 
+        cells calculated in the process.
+
+        NOTICE - the cache is stored using python lists, and therefore begins at index 0.
+        This might not the first index of the CMF, which will lead to None values in the 
+        procedding potential cells.
+        """
+        x_target, y_target = location
+        x_start, y_start = self.initial_loc
+
+        if self.potential_cache[x_target][y_target] is not None:
+            return self.potential_cache[x_target][y_target]
+
+        # TODO - move this implementation to a decorator. Something like @cache.
+
+        # We calculate the CMF potential at a location (x,y) by first multiplying the M_x
+        # matrices, and then the M_y matrices. Using this convention, we can use the cached values
+        # to make our calculation shorter.
+
+        # Proceeding the x direction from the initial location - y is constant and equal to y_start
+        for x_i in range(x_start, x_target+1):
+            if self.potential_cache[x_i][y_start] is not None:
+                continue
+
+            self.potential_cache[x_i][y_start] = \
+                self.potential_cache[x_i-1][y_start] * self.Mx.subs({x: x_i-1, y: y_start})
+
+        # Proceeding the y direction from (x_target, y_start)
+        for y_i in range(y_start, y_target+1):
+            if self.potential_cache[x_target][y_i] is not None:
+                continue
+
+            self.potential_cache[x_target][y_i] = \
+                self.potential_cache[x_target][y_i-1] * self.My.subs({x: x_target, y: y_i-1})
+
+        return self.potential_cache[x_target][y_target]
+
+    def calculate_over_range(self, max_x, max_y):
+        """
+        calculates the CMF in first quandrant up to (max_x, max_y).
+        """
+        
+        # The calculation method in `potential` takes steps over the x axis first, then over y.
+        # Since it caches elements while calculating, this will lead to caching of the entire space. 
+        for x_i in range(self.initial_loc[0], max_x+1):
+            self.potential([x_i, max_y])
+
+    def dump_potential_to_file(self, dest_path):
```

Comment:
These functions should probably be part of that new object as well. Namely, you should have something like cmf_object.cache_object.dump() \ load()